### PR TITLE
[RPA-749] sub processes lazy loading

### PR DESCRIPTION
**Pull request by star-in-the-jar**
## Description

Teraz podprocesy są pobierane dopiero po wciśnięciu przycisku "rozwiń" na wierszu z procesem-rodzicem.
image

Podczas ich pobierania wyświetla się loading bar:
image

A w przypadku niepowodzenia otrzymujemy powiadomienie w prawym dolnym rogu oraz znika przycisk "rozwiń":
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessInstanceResource.java [hunk 1]
```diff
@@ -115,4 +108,20 @@ public ResponseEntity<ProcessInstanceDTO> getProcessInstance(@PathVariable UUID
...
+
+    /**
+     * {@code GET /process-instances/{id}/subprocesses} : get all subprocesses for the processInstance of a particular ID.
+     *
+     * @param id the ID of the processInstance.
+     * @return the {@link ResponseEntity} with status {@code 200 (OK)} and the list of subprocesses in the body,
+     *         or with status {@code 404 (Not Found)} if the processInstance does not exist.
+     */
+    @GetMapping("/process-instances/{id}/subprocesses")
+    public ResponseEntity<List<ProcessInstanceDTO>> getProcessInstanceSubprocesses(@PathVariable UUID id) {
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessInstanceResource.java:121]: add an annotation for required featured keys. Use `PROCESS_INSTANCE_READ`

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.thunks.ts [hunk 2]
```diff
@@ -48,6 +48,18 @@ export const getProcessInstance = createAsyncThunk<IProcessInstance, {
...
 
+export const getSubProcesses = createAsyncThunk<IProcessInstance[], {
+    processInstanceId: string;
+}>(
+    'processInstances/getSubProcesses',
+    ({ processInstanceId }) => 
+        Axios.get<IProcessInstance[]>(`/api/process-instances/${processInstanceId}/subprocesses`)
+            .then((response) => {
+                if(response.data.length > 0) return response.data;
+                throw new Error('Could not found any sub processes');
```
Comment (star-in-the-jar) [inline runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.thunks.ts]: this is message just for debug. there is a translation

**Review by star-in-the-jar — COMMENTED** [review]
Self reviewed

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx [hunk 3]
```diff
@@ -102,6 +102,19 @@ const HistoryTable = forwardRef<any, HistoryTableProps>(({ botId, processId, sx,
         },
     }));
 
+    const handleRowExpand = (currRow) => {
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx]: is this file named `*.jsx`? xd

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx [hunk 4]
```diff
@@ -102,6 +102,19 @@ const HistoryTable = forwardRef<any, HistoryTableProps>(({ botId, processId, sx,
...
+            .then(unwrapResult)
+            .catch(() => {
+                enqueueSnackbar(
+                    translate('History.Table.Error.SubProcessesNotFound'),
+                    { variant: 'error' },
+                );
+                dispatch(processInstanceActions.updateProcessInstance({ id: currRow.original.id, hasSubProcesses: false }));
+            });
+    };
+
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx]: imo this whole logic can be in file `HistoryTable.columns.tsx` you have all information needed there just import some hooks. It's desired to move logic outside of the main components

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts [hunk 5]
```diff
@@ -31,6 +31,7 @@ export interface DataTableProps<T extends object> {
 
 export interface TableRowExpanderProps {
     row: Row;
+    handleClick?: (row: Row) => void;
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts]: export `Row` from Table.types.ts so we rely only on logic from single source. its our facade for react-table

Comment (preludium) [reply runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts]: fyi you imported Row react-table in another file

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.thunks.ts (hunk 2 above)
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.thunks.ts]: imo this logic should be with your dispatch in the component. We need to provide a translation here + If someone were to use this thunk, he/she might be surprised that the request was completed with a 200 status code but got an error

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.utils.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+    state.all.page.content = pageContent
+        .map((instance: IProcessInstance) => 
+            instance.id !== id
+                ? instance 
+                : ({ 
+                    ...instance,
+                    subProcesses: subProcesses !== undefined ? subProcesses : instance.subProcesses, 
+                    hasSubProcesses: hasSubProcesses !== undefined ? hasSubProcesses : instance.hasSubProcesses,
+                    isLoadingSubProcesses: isLoadingSubProcesses !== undefined ? isLoadingSubProcesses : instance.isLoadingSubProcesses,
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.utils.ts]: id don't see the type for `isLoadingSubProcesses` in `ProcessInstanceState` interface. Weird if you didn't get any eslint error.
1. provide field for it in ProcessInstanceState
2. keep it outside of the process_instance object there should be another object next to it

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx (hunk 3 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx]: rhetorical question? xD
added types - done

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx (hunk 4 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/HistoryTable/HistoryTable.tsx]: noted!
done

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts (hunk 5 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts]: moved all HistoryTable.tsx types to Table.types

Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts]: done

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.thunks.ts (hunk 2 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.thunks.ts]: moved logic to the dispatch.
there's translation

> what's missing in getSubProcessesResponse
yeah that proves the point. throw error removed

done

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.utils.ts (hunk 6 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.utils.ts]: I've made changes. What do you think about them?

if it's not what you expected I understand I should create new array in ProcessInstanceState storing only subProcesses related props: {id: string, isLoadingSubProcesses: boolean, hasSubProcesses: boolean, subProcesses} placed in state.all.subProcessesById?

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessInstanceResource.java (hunk 1 above)
Comment (star-in-the-jar) [reply runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessInstanceResource.java:121]: sorry I missed it

done

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.state.ts [hunk 7]
```diff
@@ -14,11 +14,17 @@ export interface ProcessInstanceState {
...
+        page: Page<InstanceExtendedWithSubProcesses> | null;
     };
 }
 
 export interface ProcessInstanceRequestCriteria extends Omit<IProcessInstance, 'process' | 'bot'> {
     botId?: number;
     processId?: number;
 }
+
+export interface InstanceExtendedWithSubProcesses extends IProcessInstance {
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.state.ts]: subprocess is a single word, no need for upper case

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts [hunk 8]
```diff
@@ -24,15 +26,48 @@ export interface DataTableProps<T extends object> {
...
+    type: string;
+    payload: InstanceExtendedWithSubProcesses[];
+    meta: {
+        arg: {
+            processInstanceId: string;
+        };
+        requestId: string;
+        requestStatus: 'fulfilled';
+    }
+}
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts]: this one does not belong to Table component. It's HistoryTable related. The same goes for all the new ones that you added here. Here should be only a generic type for the general-use Table component.

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.state.ts (hunk 7 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstance/ProcessInstance.state.ts]: as you wish, 70 occurrences changed

File: runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts (hunk 8 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts]: done

Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/components/tables/Table/Table.types.ts]: done

**Review by star-in-the-jar — COMMENTED** [review]
fast self review

**Review by preludium — APPROVED** [review]

**Pull request merged**